gtsam/base/Matrix.h

```diff
@@ -549,25 +549,52 @@ namespace boost {
   namespace serialization {
 
     // split version - sends sizes ahead
-    template<class Archive>
-    void save(Archive & ar, const gtsam::Matrix & m, unsigned int /*version*/) {
+    template<class Archive,
+             typename Scalar_,
+             int Rows_,
+             int Cols_,
+             int Ops_,
+             int MaxRows_,
+             int MaxCols_>
+    void save(Archive & ar,
+              const Eigen::Matrix<Scalar_, Rows_, Cols_, Ops_, MaxRows_, MaxCols_> & m,
+              const unsigned int /*version*/) {
       const size_t rows = m.rows(), cols = m.cols();
       ar << BOOST_SERIALIZATION_NVP(rows);
       ar << BOOST_SERIALIZATION_NVP(cols);
       ar << make_nvp("data", make_array(m.data(), m.size()));
     }
 
-    template<class Archive>
-    void load(Archive & ar, gtsam::Matrix & m, unsigned int /*version*/) {
+    template<class Archive,
+             typename Scalar_,
+             int Rows_,
+             int Cols_,
+             int Ops_,
+             int MaxRows_,
+             int MaxCols_>
+    void load(Archive & ar,
```

Comment:
This cannot work for fixed arrays. They have no resize. So, you need to either specialize the template (preferred), or do some conditional logic below. I would also (unless you want to think about al possible combinations, remove all but the Scalar, Rows, Cols template parameters.

Reply:
I did some research and I found a couple of things:

1. Fixed sized matrices support `resize` for uniformity of the API though have assertions in place to ensure the resize is not incorrect. Quoting from the Eigen docs:

> All these methods are still available on fixed-size matrices, for the sake of API uniformity. Of course, you can't actually resize a fixed-size matrix. Trying to change a fixed size to an actually different value will trigger an assertion failure;

2. I spent a lot of time researching this particular problem yesterday, since I found this slightly odd too. This [StackOverflow answer](https://stackoverflow.com/a/22903063/1236990) was my main source of understanding, and I found confidence in this answer when I saw [eigen_typekit in ROS](http://docs.ros.org/lunar/api/eigen_typekit/html/eigen__mqueue_8cpp_source.html) use the same code. This [gist](https://gist.github.com/mtao/5798888) also provides the same code, though they seem independently developed since the gist doesn't use some in-built Eigen functionality. This [other gist](https://gist.github.com/artivis/288898ff877130e9a58ed04aeb9e59b5) does partial template specialization so putting it here as a refernce.

3. I agree with keeping only Scalar, Rows and Cols. It just seemed like a good idea to cover our bases by including all parameters, especially after seeing the examples in my above comment.

Thoughts?